src/Features/CSharp/Portable/ConvertForToForEach/CSharpConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,156 @@
+                            identifierName.Identifier.ValueText == iterationVariable.ValueText &&
+                            binaryExpression.Right is MemberAccessExpressionSyntax)
+                        {
+                            memberAccess = (MemberAccessExpressionSyntax)binaryExpression.Right;
+
+                            var incrementor = forStatement.Incrementors[0];
+
+                            if (TryGetStepValue(iterationVariable, incrementor, out stepValue, cancellationToken))
+                            {
+                                return true;
+                            }
+                        }
+                    }
+                }
+            }
+
+            iterationVariable = default;
+            memberAccess = default;
+            initializer = default;
+            stepValue = default;
+            return false;
+        }
+
+        private static bool TryGetStepValue(
+            SyntaxToken iterationVariable, ExpressionSyntax incrementor,
+            out ExpressionSyntax stepValue, CancellationToken cancellationToken)
+        {
+            // support
+            //  x++
+            //  ++x
+            //  x += constant_1
+
+            ExpressionSyntax operand;
+            if (incrementor.Kind() == SyntaxKind.PostIncrementExpression)
+            {
+                operand = ((PostfixUnaryExpressionSyntax)incrementor).Operand;
+                stepValue = default;
+            }
+            else if (incrementor.Kind() == SyntaxKind.PreIncrementExpression)
+            {
+                operand = ((PrefixUnaryExpressionSyntax)incrementor).Operand;
+                stepValue = default;
+            }
+            else if (incrementor.Kind() == SyntaxKind.AddAssignmentExpression)
+            {
+                var assignment = (AssignmentExpressionSyntax)incrementor;
+                operand = assignment.Left;
+                stepValue = assignment.Right;
+            }
+            else
+            {
+                stepValue = default;
+                return false;
+            }
+
+            return operand is IdentifierNameSyntax identifierName &&
+                identifierName.Identifier.ValueText == iterationVariable.ValueText;
+        }
+
+
+        protected override SyntaxNode ConvertForNode(
+            ForStatementSyntax forStatement, TypeSyntax typeNode, 
+            SyntaxToken foreachIdentifier, ExpressionSyntax collectionExpression,
+            ITypeSymbol iterationVariableType, OptionSet optionSet)
+        {
+            if (typeNode == null)
+            {
+                // types are not apperant in foreach statements.
+                var isBuiltInTypeContext = TypeStyleHelper.IsBuiltInType(iterationVariableType);
+                if (TypeStyleHelper.IsImplicitStylePreferred(
+                        optionSet, isBuiltInTypeContext, isTypeApparentContext: false))
+                {
+                    typeNode = SyntaxFactory.IdentifierName("var");
```

Comment:
duplication

Reply:
sorry, don't know what this means.

Reply:
could you use `?:`?

Reply:
`typeNode =` is duplicated in each branch

Reply:
it feels like it would be very large and unweildy.  I found this to read better.

Reply:
we don't have a general rule of avoiding hte pattern of:

```c#
SomeVar v;

// initialize through branchign code

// use v
```

Devs decide on a case by case basis.

Reply:
#wontfix

Reply:
this is nice. I should reuse this.